src/GitVersionCore/StringFormatWith.cs

```diff
@@ -27,16 +27,44 @@ public static string FormatWith<T>(this string template, T source)
 
             // {MajorMinorPatch}+{Branch}
             var objType = source.GetType();
+            const string ENV_VAR_INDICATOR = "env:";
             foreach (Match match in TokensRegex.Matches(template))
             {
                 var memberAccessExpression = TrimBraces(match.Value);
-                Func<object, string> expression = CompileDataBinder(objType, memberAccessExpression);
-                string propertyValue = expression(source);
+                string propertyValue = null;
+
+                // Support evaluation of environment variables in the format string
+                // For example: {env:JENKINS_BUILD_NUMBER??fall-back-string}
+
+                if (memberAccessExpression.StartsWith(ENV_VAR_INDICATOR))
```

Comment:
Can't you just check for the match of `env:` in the regex instead of using this constant? With [Named Matched Subexpressions](https://docs.microsoft.com/en-us/dotnet/standard/base-types/grouping-constructs-in-regular-expressions#named-matched-subexpressions) in the regex you can reference the [match group](https://docs.microsoft.com/en-us/dotnet/api/system.text.regularexpressions.group?view=netframework-4.7.1) by name and check its [`.Success` property](https://docs.microsoft.com/en-us/dotnet/api/system.text.regularexpressions.group.success).

Reply:
Addressed in the latest commit